sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -577,55 +611,78 @@ public void setup() throws Exception {
         dataSource = spec.getDataSourceConfiguration().buildDatasource();
         connection = dataSource.getConnection();
         connection.setAutoCommit(false);
-        preparedStatement = connection.prepareStatement(spec.getStatement());
-      }
-
-      @StartBundle
-      public void startBundle() {
-        batchCount = 0;
       }
 
       @ProcessElement
       public void processElement(ProcessContext context) throws Exception {
         T record = context.element();
 
-        preparedStatement.clearParameters();
-        spec.getPreparedStatementSetter().setParameters(record, preparedStatement);
-        preparedStatement.addBatch();
-
-        batchCount++;
+        records.add(record);
 
-        if (batchCount >= spec.getBatchSize()) {
+        if (records.size() >= spec.getBatchSize()) {
           executeBatch();
         }
       }
 
+      private void processRecord(T record, PreparedStatement preparedStatement) {
+        try {
+          preparedStatement.clearParameters();
+          spec.getPreparedStatementSetter().setParameters(record, preparedStatement);
+          preparedStatement.addBatch();
+        } catch (Exception e) {
+          throw new RuntimeException(e);
+        }
+      }
+
       @FinishBundle
       public void finishBundle() throws Exception {
         executeBatch();
       }
 
-      private void executeBatch() throws SQLException {
-        if (batchCount > 0) {
-          preparedStatement.executeBatch();
-          connection.commit();
-          batchCount = 0;
+      private void executeBatch() throws SQLException, IOException, InterruptedException {
+        if (records.size() == 0) {
+          return;
+        }
+        Sleeper sleeper = Sleeper.DEFAULT;
+        BackOff backoff = BUNDLE_WRITE_BACKOFF.backoff();
+        while (true) {
+          PreparedStatement preparedStatement = connection.prepareStatement(spec.getStatement());
+          try {
+            // add each record in the statement batch
+            for (T record : records) {
+              processRecord(record, preparedStatement);
+            }
+            // execute the batch
+            preparedStatement.executeBatch();
+            // commit the changes
+            connection.commit();
+            break;
+          } catch (SQLException exception) {
+            if (!exception.getSQLState().equals(spec.getLockSqlState())) {
+              throw exception;
+            }
+            LOG.warn("Lock detected, retrying", exception);
+            // clean up the statement batch and the connection state
+            preparedStatement.clearBatch();
+            connection.rollback();
```

Comment:
Is `rollback()` always valid in this case? What if the database auto-rollbacks deadlocked transactions: is it valid to call `rollback()` again even though there is no currently active transaction? Or is there always a default one?